cynet/client.go

```diff
@@ -0,0 +1,582 @@
+			a.conf.ClientOptions.OnWarning("timeout waiting for fetch loop to exit; proceeding with cleanup")
+			a.fetchOnce.Do(func() { close(a.chFetchLoop) })
+		}
+		err1 = a.uspClient.Drain(1 * time.Minute)
+		_, err2 = a.uspClient.Close()
+		a.httpClient.CloseIdleConnections()
+		close(a.chStopped)
+	})
+	if err1 != nil {
+		return err1
+	}
+	return err2
+}
+
+func (a *CynetAdapter) authenticate() error {
+	// a.conf.ClientOptions.DebugLog("authenticating with Cynet API")
+
+	authURL := fmt.Sprintf("%s%s", a.conf.URL, authEndpoint)
+
+	var authReq AuthRequest
+
+	// If both AccessKey/SecretKey and User/Password pairs are provided, default to AccessKey/SecretKey
+	if a.conf.AccessKey != "" && a.conf.SecretKey != "" {
+		authReq = AuthRequest{
+			User:     a.conf.AccessKey,
+			Password: a.conf.SecretKey,
+		}
+	} else {
+		authReq = AuthRequest{
+			User:     a.conf.User,
+			Password: a.conf.Password,
+		}
+	}
+
+	jsonData, err := json.Marshal(authReq)
+	if err != nil {
+		return fmt.Errorf("failed to marshal auth request: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(a.ctx, 30*time.Second)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, "POST", authURL, bytes.NewBuffer(jsonData))
+	if err != nil {
+		return fmt.Errorf("failed to create auth request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := a.httpClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("auth request failed: %w", err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read auth response: %w", err)
+	}
+
+	if resp.StatusCode == http.StatusUnauthorized {
+		return fmt.Errorf("authentication failed: invalid credentials (401 Unauthorized)")
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("authentication failed with status %d: %s", resp.StatusCode, string(body))
+	}
+
+	var authResp AuthResponse
+	if err := json.Unmarshal(body, &authResp); err != nil {
+		return fmt.Errorf("failed to unmarshal auth response: %w", err)
+	}
+
+	if authResp.AccessToken == "" {
+		return fmt.Errorf("received empty access token from auth endpoint")
+	}
+
+	a.accessToken = authResp.AccessToken
+
+	return nil
+}
+
+func (a *CynetAdapter) fetchEvents() {
+	defer func() {
+		a.fetchOnce.Do(func() { close(a.chFetchLoop) })
+	}()
+
+	ticker := time.NewTicker(queryInterval * time.Second)
+	defer ticker.Stop()
+
+	// Initial fetch
+	a.runFetchCycle()
+
+	for {
+		select {
+		case <-a.ctx.Done():
+			return
+		case <-ticker.C:
+			a.runFetchCycle()
+		}
+	}
+}
+
+func (a *CynetAdapter) runFetchCycle() {
+	// a.conf.ClientOptions.DebugLog("starting fetch cycle")
+
+	// Fetch alerts
+	alerts, err := a.fetchAlerts()
+	if err != nil {
+		a.conf.ClientOptions.OnError(fmt.Errorf("failed to fetch alerts: %w", err))
+		a.consecutiveErrors++
+		if a.consecutiveErrors >= maxConsecutiveErrors {
```

Comment:
`cleanupDedupe()` is called on every page iteration. Should be called once per fetch cycle, before the pagination loop starts.